jena-arq/src/main/java/org/apache/jena/sparql/resultset/CSVOutput.java

```diff
@@ -102,6 +102,62 @@ public void format(OutputStream out, ResultSet resultSet)
         }
     }
 
+    @Override
+    public void format(OutputStream out, ResultSet resultSet, StringBuilder cacheBuilder)
+    {
+        try {
+            Writer w = FileUtils.asUTF8(out) ;
+            NodeToLabelMap bnodes = new NodeToLabelMap() ;
+            w = new BufferedWriter(w) ;
+
+            String sep = null ;
+            List<String> varNames = resultSet.getResultVars() ;
+            List<Var> vars = new ArrayList<>(varNames.size()) ;
+
+            // Convert to Vars and output the header line.
+            for( String v : varNames )
+            {
+                if ( sep != null ){
+                    w.write(sep) ;
+                    cacheBuilder.append(sep);
+                }
+                else
+                    sep = "," ;
+                w.write(csvSafe(v)) ;
+                cacheBuilder.append(csvSafe(v));
+                vars.add(Var.alloc(v)) ;
+            }
+            w.write(NL) ;
+            cacheBuilder.append(NL);
+
+            // Data output
+            for ( ; resultSet.hasNext() ; )
```

Comment:
These two loops could be managed a bit more tightly between `resultSet::forEachRemaining` and `vars.stream()::map`.